presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java

```diff
@@ -371,17 +379,67 @@ protected Node visitShowPartitions(ShowPartitions showPartitions, Void context)
             TableHandle tableHandle = metadata.getTableHandle(session, table)
                     .orElseThrow(() -> new SemanticException(MISSING_TABLE, showPartitions, "Table '%s' does not exist", table));
 
-            TableLayout layout = getLayout(showPartitions, showPartitions.getTable(), showPartitions.getWhere())
-                    .orElseThrow(() -> new SemanticException(NOT_SUPPORTED, showPartitions, "Could not find layout for table: %s", table));
+            Optional<TableLayout> layout = getLayout(showPartitions, showPartitions.getTable(), showPartitions.getWhere());
+            List<Expression> rows = getPartitionsAsValuesRows(showPartitions, table, layout);
+            List<String> partitioningColumns = getPartitioningColumns(showPartitions, table, tableHandle, layout);
 
+            Optional<Expression> where;
+            if (rows.isEmpty()) {
+                // VALUES does not allow no rows
+                List<Expression> row = partitioningColumns.stream()
+                        .map(partitioningColumn -> new NullLiteral())
+                        .collect(toImmutableList());
+                rows = ImmutableList.of(new Row(row));
+                where = Optional.of(FALSE_LITERAL);
+            }
+            else {
+                where = showPartitions.getWhere();
+            }
+
+            return simpleQuery(
+                    selectList(partitioningColumns.stream().map(QueryUtil::identifier).collect(toImmutableList())),
+                    aliased(new Values(rows), "x", partitioningColumns),
+                    where,
+                    Optional.empty(),
+                    Optional.empty(),
+                    orderBy(showPartitions.getOrderBy()),
+                    showPartitions.getLimit());
+        }
+
+        private List<String> getPartitioningColumns(
+                ShowPartitions showPartitions,
+                QualifiedObjectName table,
+                TableHandle tableHandle,
+                Optional<TableLayout> layoutFromQuery)
+        {
             Map<ColumnHandle, String> columnHandleNames = ImmutableBiMap.copyOf(metadata.getColumnHandles(session, tableHandle)).inverse();
+            TableLayout tableLayout;
+            if (layoutFromQuery.isPresent()) {
+                tableLayout = layoutFromQuery.get();
+            }
+            else {
+                List<TableLayoutResult> layouts = metadata.getLayouts(session, tableHandle, new Constraint(TupleDomain.all(), alwaysFalse()), Optional.empty());
```

Comment:
@findepi Here comes the magic `new Constraint(TupleDomain.all(), alwaysFalse())` it reads layouts without any partition loaded.

Reply:
👍  Clearly, this remains a hack, since Hive connector could return no layouts, signalling effectively empty table. It doesn't do that -- it's their problem.